### PR TITLE
#54: Make code more extensible

### LINKED ISSUE #54: Improve ModelServerLauncher for easier customization
Currently the model server launcher provides some methods to add customizations, e.g., `addEPackageConfigurations`. However, in other cases no such method exists, e.g., when trying to add additional routing, see [1] and [2]. I believe the API could be improved in this regard to allow easier integration of additional modules (generics in the return type of `getModules` make it unnecessarily difficult) and to already provide hooks for classes which may be implemented more than once, i.e., that have a collection binding.

I believe we could discuss whether that improved customizations should be part of the `ModelServerModule` itself or the `ModelServerLauncher`.

[1] https://spectrum.chat/emfcloud/modelserver/model-server-api-to-return-models-of-a-specific-type~b4a27e5d-48e6-41fe-8177-f66056f8c77f
[2] https://github.com/eclipse-emfcloud/emfcloud-modelserver/commit/954ffed79b78b772b9b657a80e06f88f1ded9575

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Introduce interfaces for services and bind implementation in module
-- Provide methods in module for users to bind their own implementation
-- Provide methods in module for users to adapt multi-bindings

- Use protected methods and fields in implementations

- Code refactorings
-- Extract context request and response method in utility classes
-- Provide new UriHelper class to deal with URIs
-- Remove cascading ifPresentOrElse-blocks

- Rename and adapt test cases where necessary
- Fix p2 target update issue created by auto-formatting

- Add showcase for custom routing by supporting a 'counter' REST call

Signed-off-by: Martin Fleck <mfleck@eclipsesource.com>

### CODE REVIEW

**Review by ndoschek — CHANGES_REQUESTED** [review]
Thank you very much for this great change Martin! :star_struck: The extensibility is big step forward and the code cleanup is extraordinary!

I tested the standalone version and consumed the eclipse product in another project. Everything still seems to work as expected as far as I could see :100: 

However, I found a few minor issues (please see inline).
Other than that, I noticed that the dependency `com.fasterxml.jackson.core.jackson-databind` is not available anymore if the feature group is consumed via p2 plugins. So I had to add the dependency additionally, which is no big problem, but maybe we can fix that.

File: examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/CoffeeRouting.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      this.javalin = javalin;
+   }
+
+   @Override
+   public void bindRoutes() {
+      javalin.routes(this::endpoints);
+   }
+
+   protected void endpoints() {
+      path("api/v1", this::coffeeEndpoints);
```
Comment (ndoschek) [inline examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/CoffeeRouting.java:53]:
```suggestion
      path(ModelServerPathsV1.BASE_PATH, this::coffeeEndpoints);
```

File: bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/configuration/DefaultServerConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+import com.google.inject.Inject;
+
+/**
+ * Singleton class that holds configurations parameters for a model server
+ * instance.
+ *
+ */
+public class DefaultServerConfiguration implements ServerConfiguration {
+   private static Logger LOG = Logger.getLogger(DefaultServerConfiguration.class);
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply bundles/org.eclipse.emfcloud.modelserver.emf/src/org/eclipse/emfcloud/modelserver/emf/configuration/DefaultServerConfiguration.java]: Good catch, I fixed it!

File: examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/ExampleServerLauncher.java [hunk 3]
```diff
@@ -49,34 +46,38 @@
...
 
-   public static void main(String[] args) throws ParseException {
-      ModelServerLauncher.configureLogger();
-      CLIParser.create(args, CLIParser.getDefaultCLIOptions(), PROCESS_NAME);
+   public static void main(final String[] args) throws ParseException {
+      final CLIBasedModelServerLauncher launcher = new CLIBasedModelServerLauncher(
+         createCLIParser(args),
+         new ExampleServerModule());
+      launcher.run();
+   }
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/ExampleServerLauncher.java:54]: You are right, I made a mistake in my refactoring thinking that we always set the UI Schema folder but in fact it is only set if no workspace is given and we create a temp workspace. I fixed the issue now, thank you!

File: examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/ExampleServerLauncher.java [hunk 4]
```diff
@@ -49,34 +46,38 @@
...
+
+   protected static CLIParser createCLIParser(final String[] args) throws ParseException {
+      CLIParser parser = new CLIParser(args, CLIParser.getDefaultCLIOptions(), PROCESS_NAME, 8081);
+      ensureWorkspaceRoot(parser);
+      ensureUISchemaFolder(parser);
+      return parser;
+   }
+
+   protected static void ensureUISchemaFolder(final CLIParser parser) throws ParseException {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/ExampleServerLauncher.java]: See my previous comment, I made a mistake in the refactoring. Thank you very much for catching that!

File: examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/CoffeeRouting.java (hunk 1 above)
Comment (martin-fleck-at) [reply examples/org.eclipse.emfcloud.modelserver.example/src/main/java/org/eclipse/emfcloud/modelserver/example/CoffeeRouting.java]: Ah, yes, constants, we need to provide AND use them ;-) Thanks!

**Comment by martin-fleck-at** [comment]
@ndoschek Thank you for your thorough review, I think I fixed it all up in a follow-up commit.

**Comment by ndoschek** [comment]
@martin-fleck-at Thank you for the quick update Martin!
However, I noticed an issue: If a workspaceRoot is set on startup, the initialization of this folder does not work anymore.
If I set a breakpoint at `DefaultModelResourceManager::initialize#75`, I can see that `getWorkspaceRootURI` delivers an empty string. The method `setWorkspaceRoot` of `ServerConfiguration` is called afterwards, which is too late in this case.

Other than that, everything looks great to me! :+1:

**Comment by martin-fleck-at** [comment]
@ndoschek Ah, you are right, I am very sorry for that! It seems like the magic of DI got me, I now ensure that we configure the server configuration BEFORE doing anything else in our launcher. Thank you very much for catching that!

**Review by ndoschek — APPROVED** [review]
Thank you very much Martin, everything works great now! :+1:

**Pull request merged**